### PR TITLE
Optimize `decode_len`

### PR DESCRIPTION
**Pull request by bkchr**
Instead of reading the full storage value into the runtime, we only read
at maximum `5bytes` from the storage into the runtime. Furthermore this
drops any handling with regards to set default values in
`decl_storage!`. If the value does not exists or the decoding of the
length fails, it will return `None`. To prevent people from messing
stuff up, this feature relies on the `StorageDecodeLength` trait that is
sealed by `frame-support` (aka only implementable inside this crate).

### CODE REVIEW

**Comment by gavofyork** [comment]
guess a master merge is needed..

File: frame/support/src/storage/mod.rs [hunk 1]
```diff
@@ -493,7 +518,29 @@ pub trait StoragePrefixedMap<Value: FullCodec> {
...
+/// and that the length is a `Compact<u32>`.
+///
+/// This trait is sealed.
+pub trait StorageDecodeLength: private::Sealed + codec::DecodeLength {
+	/// Decode the length of the storage value at `key`.
+	///
+	/// This function assumes that the length is at the beginning of the encoded object
+	/// and is a `Compact<u32>`.
+	///
+	/// Returns `None` if the storage value does not exists or the decoding failed.
```
Comment (apopiak) [inline frame/support/src/storage/mod.rs:532]:
```suggestion
	/// Returns `None` if the storage value does not exist or the decoding failed.
```

**Comment by shawntabrizi** [comment]
> 	/// This function assumes that the length is at the beginning of the encoded object and is a `Compact<u32>`.

When is this true?

I also see we lost the ability to `decode_len` on a vector?

**Comment by bkchr** [comment]
This is true for every container type we support.

And where did we lost decode_len for vectors?

**Comment by shawntabrizi** [comment]
my mistake, it was on vectors with default value, and you already mentioned we lost the ability to get length where no value exists (default)

**Pull request merged**